### PR TITLE
Feature/fcm generic connector

### PR DESCRIPTION
**Pull request by Bogh**

### CODE REVIEW

**Comment by cosminrentea** [comment]
solved conflict with master and pushed.
there are still some compilation errors in integration tests.

File: server/connector/queue.go [hunk 1]
```diff
@@ -62,11 +73,23 @@ func (q *queue) worker(i int) {
 				}).Error("Error handling connector response")
 			}
 		}
-		q.wg.Done()
+		defer q.wg.Done()
```
Comment (cosminrentea) [inline server/connector/queue.go]: should this be simply q.wg.Done() , without `defer` ?

File: server/fcm/fcm.go [hunk 2]
```diff
@@ -43,408 +25,110 @@ type Config struct {
...
+	// sender connector.Sender
+}
+
+// New creates a new *fcm and returns it as an connector.ReactiveConnector
+func New(router router.Router, sender connector.Sender, config Config) (connector.ReactiveConnector, error) {
+	baseConn, err := connector.NewConnector(router, sender, connector.Config{
+		Name:       "fcm",
+		Schema:     schema,
+		Prefix:     *config.Prefix,
+		URLPattern: fmt.Sprintf("/{device_token}/{user_id}/{%s:.*}", connector.TopicParam),
```
Comment (cosminrentea) [inline server/fcm/fcm.go]: please extract `device_token` and `user_id` as constants
reuse them in this package (see for instance line 129, below, maybe some other places)

File: server/fcm/fcm_test.go [hunk 3]
```diff
@@ -1,468 +1,143 @@
...
-
-	// then: all expected subscriptions were called
-	a.Equal(0, len(testRoutes))
-}
-
-func TestRemoveTrailingSlash(t *testing.T) {
-	assert.Equal(t, "/foo", removeTrailingSlash("/foo/"))
-	assert.Equal(t, "/foo", removeTrailingSlash("/foo"))
-}
+// func TestConnector_Check(t *testing.T) {
```
Comment (cosminrentea) [inline server/fcm/fcm_test.go]: this should be un-commented if possible, and the TODO removed
now the Check func is public again, needs a test
if Check is not really working now, there is a task I created earlier on JIRA -> healthchecks for apns+fcm

File: server/fcm_integration_test.go [hunk 4]
```diff
@@ -99,18 +101,24 @@ func TestFCMRestart(t *testing.T) {
...
+
 	time.Sleep(50 * time.Millisecond)
 	testutil.ResetDefaultRegistryHealthCheck()
 	a.NoError(s.Start())
 
 	//TODO Cosmin Bogdan add 2 calls to assertMetrics before and after the next block
 
 	// read the other 2 messages
-	for i := 0; i < 2; i++ {
+	for i := 0; i < 1; i++ {
```
Comment (cosminrentea) [inline server/fcm_integration_test.go]: funny :)

File: testutil/testutil.go [hunk 5]
```diff
@@ -83,109 +80,6 @@ func ResetDefaultRegistryHealthCheck() {
...
-	}).Debug("Create FCM sender")
-	httpClient := &http.Client{Transport: rt}
-
-	client := gcm.NewSender("1234", 0, 3*time.Second)
-	client.HTTPClient = httpClient
-
-	return client
-}
-
-func CreateRoundTripperWithJsonResponse(statusCode int, body string, doneC chan bool) RoundTripperFunc {
```
Comment (cosminrentea) [inline testutil/testutil.go]: hooray! interfaces FTW, instead of these funcs :)

File: server/connector/subscriber.go [hunk 6]
```diff
@@ -153,8 +159,16 @@ func GenerateKey(topic string, params map[string]string) string {
...
 	io.WriteString(h, topic)
-	for k, v := range params {
-		io.WriteString(h, fmt.Sprintf("%s:%s", k, v))
+
+	// compute the hash with ordered params keys
+	keys := make([]string, 0, len(params))
+	for k := range params {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
```
Comment (cosminrentea) [inline server/connector/subscriber.go]: I guess this was the reason of the random unit-test-failures ?

File: server/connector/queue.go (hunk 1 above)
Comment (Bogh) [inline server/connector/queue.go]: It's a mistake. I changed stuff during tests to find some issue and forgot it.

**Pull request merged**